### PR TITLE
Support sparse registry references in Lockfiles

### LINKED ISSUE #604: `cargo audit` doesn't work well with `sparse-registry`
[Databend](https://github.com/datafuselabs/databend) start a PR https://github.com/datafuselabs/databend/pull/6437 in insponse to [Call for testing: Cargo sparse-registry](https://internals.rust-lang.org/t/call-for-testing-cargo-sparse-registry/16862).

During this test, we found `cargo audit` doesn't work well with `sparse-registry`:

I expect to see this index updated within few seconds, however:

Before PR (18s):

https://github.com/datafuselabs/databend/runs/7169319361?check_suite_focus=true#step:3:1457

```shell
Sun, 03 Jul 2022 14:35:06 GMT Run cargo audit --db ./target/advisory-db
Sun, 03 Jul 2022 14:35:06 GMT Fetching advisory database from `[https://github.com/RustSec/advisory-db.git`](https://github.com/RustSec/advisory-db.git%60)
Sun, 03 Jul 2022 14:35:07 GMT Loaded 419 security advisories (from ./target/advisory-db)
Sun, 03 Jul 2022 14:35:07 GMT Updating crates.io index
Sun, 03 Jul 2022 14:35:25 GMT Scanning Cargo.lock for vulnerabilities (691 crate dependencies)
```

Within PR (sparse-registry disabled) (19s):

https://github.com/datafuselabs/databend/runs/7181217966?check_suite_focus=true#step:3:1462

```shell
Mon, 04 Jul 2022 13:27:26 GMT Run cargo audit --db ./target/advisory-db
Mon, 04 Jul 2022 13:27:26 GMT Fetching advisory database from `[https://github.com/RustSec/advisory-db.git`](https://github.com/RustSec/advisory-db.git%60)
Mon, 04 Jul 2022 13:27:27 GMT Loaded 419 security advisories (from ./target/advisory-db)
Mon, 04 Jul 2022 13:27:27 GMT Updating crates.io index
Mon, 04 Jul 2022 13:27:46 GMT Scanning Cargo.lock for vulnerabilities (693 crate dependencies)
```

Within PR (sparse-registry enabled) (18s):

https://github.com/datafuselabs/databend/runs/7181884244?check_suite_focus=true#step:3:1462

```shell
Mon, 04 Jul 2022 14:09:57 GMT Run cargo -Z sparse-registry audit --db ./target/advisory-db
Mon, 04 Jul 2022 14:09:57 GMT Fetching advisory database from `[https://github.com/RustSec/advisory-db.git`](https://github.com/RustSec/advisory-db.git%60)
Mon, 04 Jul 2022 14:09:58 GMT Loaded 419 security advisories (from ./target/advisory-db)
Mon, 04 Jul 2022 14:09:58 GMT Updating crates.io index
Mon, 04 Jul 2022 14:10:16 GMT Scanning Cargo.lock for vulnerabilities (693 crate dependencies)
```

### PR DESCRIPTION
**Pull request by rbtcollins**
Closes #604

Signed-off-by: Robert Collins <robert.collins@cognite.com>

### CODE REVIEW

**Comment by rbtcollins** [comment]
Note that this doesn't on its own fix the cargo-audit bug with `sparse+` references in lockfiles because of the non-local use of cargo-lock by cargo-audit: a release of cargo-lock and then an update to use the released version will be required to fully fix the bug.

**Comment by rbtcollins** [comment]
Looking more closely at #604 it is possible that it is describing a different problem : the one this patch solves is explicitly annotated sparse registries, such as happens when a company or organisation has an internal registry for internal crates, and that internal registry happens to be sparse.

**Comment by tarcieri** [comment]
`source.rs` is effectively vendored from what is now https://github.com/rust-lang/cargo/blob/master/src/cargo/core/source/source_id.rs

Making any changes which diverge from upstream does not make sense to me.

**Comment by rbtcollins** [comment]
Ok. Well I can redo the change to be signed with cargo's own form. Is the vendoring automated, or just a copy+fix-up until it builds?

**Comment by tarcieri** [comment]
It’s manually vendored

**Comment by rbtcollins** [comment]
I've changed the change to match the Cargo expression of the logic.

**Comment by rbtcollins** [comment]
@tarcieri ping - I think this meets your needs vis-a-vis not diverging now

**Review by tarcieri — APPROVED** [review]

File: cargo-lock/src/package/source.rs [hunk 1]
```diff
@@ -109,9 +112,14 @@ impl SourceId {
...
             }
+            "sparse" => {
+                let url = url.into_url()?;
+                Ok(SourceId::new(SourceKind::SparseRegistry, url)?
+                    .with_precise(Some("locked".to_string())))
+            }
             "path" => Self::new(SourceKind::Path, url.into_url()?),
             kind => Err(Error::Parse(format!(
-                "unsupported source protocol: {}",
+                "unsupported source protocol: `{}` from `{string}`",
```
Comment (tarcieri) [inline cargo-lock/src/package/source.rs:122]: These interpolations aren't supported on the current MSRV, which is 1.56.

**Comment by tarcieri** [comment]
Since these enums aren't marked `#[non_exhaustive]`, this is a breaking change which necessitates a major version bump (something I can take care of).

We could also bump MSRV as part of it in order to better retain compatibility with the upstream code.

**Comment by rbtcollins** [comment]
I'm happy to push up a change to be 1.56 compatible, and add `#[non_exhaustive]` to save future compat friction; or I can push up one bumping the MSRV; let me know.

**Comment by tarcieri** [comment]
Let's go ahead and bump MSRV to stay compatible with the upstream code

**Comment by rbtcollins** [comment]
MSRV bump pushed

File: cargo-lock/src/package/source.rs [hunk 2]
```diff
@@ -130,7 +138,7 @@ impl SourceId {
         Self::new(SourceKind::Git(reference), url.clone())
     }
 
-    /// Creates a SourceId from a registry URL.
+    /// Creates a SourceId from a (git) registry URL.
```
Comment (tarcieri) [inline cargo-lock/src/package/source.rs:141]: What's up with this?

**Comment by tarcieri** [comment]
Looks like the clippy MSRV needs to be bumped

File: cargo-lock/src/package/source.rs (hunk 2 above)
Comment (rbtcollins) [reply cargo-lock/src/package/source.rs]: Oh, that was a hangover from my first patch. Have synced with the cargo source now.

**Comment by rbtcollins** [comment]
Have bumped MSRV pervasively now.

**Comment by tarcieri** [comment]
Looks good now, thanks!

**Review by tarcieri — APPROVED** [review]

**Pull request merged**